### DISCUSSION TITLE
posts/postgres-is-a-graph-database

### DISCUSSION THREAD

**Discussion by giscus** [ganderzz/ganderzz.github.io#104] [Announcements]
# posts/postgres-is-a-graph-database

PostgreSQL (Postgres), is a powerful relational database that can store a wide range of data types and data structures.
When it comes to storing graph data structures we might reach for a database marketed for that use case like Neo4J or Dgraph. Hold your horses!
While Postgres is not generally thou...

https://www.dylanpaulus.com/posts/postgres-is-a-graph-database

**Comment by cparis** [comment]
Your solution is the basis for a good graph query.   Having implemented this on a large scale, you may want to do a few minor changes.

The biggest is is if you have a circular graph.  Your query will infinitely loop.   As part of the top CTE, you should add a WHERE clause to detect if the node is already in the query graph.  Usually a NOT EXIST with your edges.next_node and friend_of_friend.next_node.

This will prevent depth issues that you will end up running into.   Hope this helps.

**Reply by ambiso** [reply]
> The biggest is is if you have a circular graph. Your query will infinitely loop.

This is not true. `WITH RECURSIVE` will only not terminate if you keep adding new distinct rows. In fact the example recursive query terminates just fine with a loop in the graph.

Excerpt from the documentation:

> 
> 2. **So long as the working table is not empty**, repeat these steps:
> 
>    a. Evaluate the recursive term, substituting the current contents of the working table for the recursive self-reference. For UNION (but not UNION ALL), **discard duplicate rows and rows that duplicate any previous result row. Include all remaining rows in the result of the recursive query, and also place them in a temporary intermediate table**.
> 
>    b. **Replace the contents of the working table with the contents of the intermediate table**, then empty the intermediate table.
> 
> 

https://www.postgresql.org/docs/current/queries-with.html

**Reply by ganderzz** [reply]
Definitely! I wanted to keep this post digestible from the start, and there are so many directions to take this. 
We can also add constraints to prevent creating loops on insert/update/deletion--especially if we're dealing with [DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph) data.

**Comment by Nezteb** [comment]
It might be worth noting for readers that some other options for storing graph/tree data are:
- [Label Trees](https://www.postgresql.org/docs/current/ltree.html)
- [Adjacency Lists](https://schinckel.net/2014/11/27/postgres-tree-shootout-part-2%3A-adjacency-list-using-ctes/)
- [Closure Tables](https://lnagle.github.io/closure-table-pattern.html)

**Reply by ganderzz** [reply]
Yes! There is also the [Adjacency Matrix](https://www.programiz.com/dsa/graph-adjacency-matrix) which isn't as great for storing in a database, but gives an interesting way to manage and work with graphs.

**Comment by sgunadhya** [comment]
[Apache AGE](https://age.apache.org/) is a PostgreSQL extension that supports Graph Database operations.

**Reply by ashwinstar** [reply]
Do we have pointers on how native support mentioned in blog compares to Apache AGE?

**Comment by dariodariodario** [comment]
We had some old code using recursive queries on a graph in postgres where I work now. Despite di obvious plus of having the recursion inside postgres itself, we found the queries beaming very problematic as soon as the complexity went beyond the basic example. Resulting in looping queries in production and hard to maintain code in general. For now we have given up performance and resorted to multiple queries keeping the recursion in the application itself.

**Comment by yiannis-had** [comment]
This query:

```
WITH RECURSIVE friend_of_friend AS (
  SELECT edges.next_node
  FROM edges
  WHERE edges.previous_node = 1
  UNION
  SELECT edges.next_node
  FROM edges
  JOIN friend_of_friend ON edges.previous_node = friend_of_friend.next_node
)
SELECT nodes.data
FROM nodes
JOIN friend_of_friend ON nodes.id = friend_of_friend.next_node;
```


should give back

| **data**  |
| ------ |
| Hank  |
| Jeff    |
| Sally   |
| Sue    |

**Comment by kevinswarner** [comment]
Using this approach, how would you support ordering of nodes. For example, let's say you are building a tree where the order of child nodes is important, and that order is defined not by some property like date, but a specific insertion order for example. Would you simply add an "index" property to the edge record?
